rfcs/003-strings.md

```diff
@@ -0,0 +1,325 @@
+# Strings in windows-rs
+
+## Summary
+
+Rust and Windows each have a fair amount of string types. This RFC proposes a system for making it easy and obvious how to make the string types from both environments work well together.
+
+## Motivation
+
+Making Rust and Windows string types work together is integral to ensuring correct, performant, and boiler-plate free code.
+
+### What we want to accomplish
+
+In general, we are trying to accomplish two goals:
+
+* Correctly model the Windows string types in a way that leverages Rust's type and ownership systems to prevent incorrect use.
+* Allow for easy and low cost ways to convert between Windows and Rust types.
+
+On top of this, we want these APIs to feel natural and obvious to Rust developers.
+
+## Explanation
+
+### The Different Types of Strings
+
+#### Windows
+
+Windows has the following string types:
```

Comment:
Another way to look at this is that Windows has a variety of string types that need to be supported for interop with various APIs but that languages can and should only use them on the ABI boundaries as much as is practical. The extreme example of this is C# where only `System.String` exists in C# and all the different string types are marshaled away behind the scenes. While I'm not suggesting we go quite that far, we should try to think of these as interop types needed only for the ABI boundary and put our energy into making it as efficient and seamless as possible to go from and get back to native and idiomatic Rust string types, and not try to make Windows replacements with all the bells and whistles of Rust's string types.

Reply:
For `PCSTR`, we can just use `*const c_char` directly, and `CStr` provides `from_ptr` and `as_ptr` methods to convert to and from it. For `PSTR`, we would need `CString` and `CStr` to allow mutability.

For `PWSTR` and `PCWSTR`, it would need roughly the same, except with wide characters. `PCWSTR` (I think) should be able to be created from a `BSTR` or `HSTRING`.

Reply:
That would be the ideal situation.

However, the unfortunate situation with Rust is, that Rust does not provide a zero-overhead string type that's compatible with Windows' native string encoding. `Vec<u16>`/`&[u16]` is the best Rust has to offer in this respect. I don't see any way *"to go from and get back to native and idiomatic Rust string types"* at all.

Being pragmatic here, and assuming that unmatched surrogate pairs won't appear in practice is a dangerous route to go down. Doing so will open up code based on the `windows` crate to attacks, even if only DoS attacks.

Reply:
Indeed that's the ideal hence the "as much as is practical". So we need a null terminated wide char but we need such things as little as humanly possible with the emphasis on interop rather than string features. And since we need HSTRING anyway and HSTRING is a null terminated wide char string that should cover those needs. Even if we end up calling it CWString, for symmetry, that just happens to be implemented as an HSTRING we can at least minimize the number of implementations we're carrying around.

Reply:
That makes sense. I am a bit skeptical that we'll actually save much code by using an `HSTRING` as a backing implementation detail, but if this is truly just an implementation detail then it doesn't really matter. We can keep playing with the implementation without impacting users.